src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1412,44 +1421,97 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
+                else if (currentArg is RangeExprNode rgnNode)
+                {
+                    Type t;
+                    if (!astTypeInfoMap.TryGetValue(rgnNode.ID, out t))
                     {
-                        var argRank = GetRank(t);
-                        var paramRank = GetRank(p);
-                        if (argRank != paramRank) return false;
-
-                        // If both have rank 0, it could also be because their type info is ambiguous.
-                        if (argRank == 0 && paramRank == 0) return false;
+                        throw new Exception("unkown ast type");
                     }
+                    if (!DoesParamRankMatchArg(fep,p, t,argIndex)) return false;
                 }
                 else return false;
                 argIndex++;
             }
             return true;
         }
 
+        /// <summary>
+        /// Check that a parameter's rank and type are compatible with the given argument.
+        /// This function besides checking rank directly, also does checks using the FEP attributes.
+        /// </summary>
+        /// <param name="fep">the function end point wrapper</param>
+        /// <param name="paramType">type of param on the fep</param>
+        /// <param name="argType">actual argument type</param>
+        /// <param name="paramIndex">parameter index in list of params to fep.</param>
+        /// <returns></returns>
+        private bool DoesParamRankMatchArg(ProtoCore.CLRFunctionEndPoint fep,Type paramType, Type argType, int paramIndex)
+        {
+            if (!paramType.Equals(argType))
+            {
+                //if both types are value types, let's
+                //let coercion figure this out.
+                if (paramType.IsValueType && argType.IsValueType)
+                {
+                    return true;
+                }
+
+                var argRank = GetRank(argType);
+                var paramRank = GetParamRank(fep,paramType, paramIndex);
+
+                //if the paramRank is an arbitrary rank array
+                //replication should not occur for this parameter
+                //(except in case of rep guides, but that has been ruled out previously).
+                if (paramRank == DSASM.Constants.kArbitraryRank)
+                {
+                    return true;
+                }
+                if (argRank != paramRank) return false;
+
+                // If both have rank 0, it could also be because their type info is ambiguous.
+                // or it could be fine!
+                if (argRank == 0 && paramRank == 0)
+                {
+                    EmitILComment("param and arg are both rank 0, type info may be ambiguous.");
+                    if (argType == typeof(DSASM.CLRStackValue))
+                    {
+                        EmitILComment("argument is from a replicate call, and param is rank 0, fallback to replication.");
+                        return false;
+                    }
```

Comment:
What are the cases, outside of this `if`? Is it the case where both arg and param are reference types and have rank 0? If so, shouldn't we let replication handle fitting the arg to the param type in this case?

Reply:
not necessarily, either the param or the arg could also be value types. 
I cannot recall which, but making this change allowed some tests start passing.

I think until we find a case where this incorrectly calls a function we should allow this - we may even want to emit a try catch and use it to fallback to replication at runtime.